src/storage/gcp/GCS.h

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2017-present varchar.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <mutex>
+
+#include "common/Errors.h"
+#include "google/cloud/storage/client.h"
+#include "storage/NFileSystem.h"
+
+/**
+ * A wrapper for interacting with GCS
+ */
+namespace nebula {
+namespace storage {
+namespace gcp {
+
+class GCS : public NFileSystem {
+  static inline google::cloud::storage::ClientOptions Options() {
+    auto opts = google::cloud::storage::ClientOptions::CreateDefaultClientOptions();
+    return *opts;
+  }
+
+public:
+  GCS(const std::string& bucket)
+    : bucket_(bucket),
+      client_{ Options(), google::cloud::storage::LimitedErrorCountRetryPolicy(3) } {
```

Comment:
We can create a client in the file system class, and one nebula process only need to hold one gcs client, because it is thread safe.
There are several ways of creating the client.
Recommend #1.
1. StatusOr<gcs::Client> client = gcs::Client::CreateDefaultClient();   // use the default Google Application Default
 Credentials (ADCs)
2. auto creds = gcs::oauth2::GoogleDefaultCredentials();
    gcs::Client client(gcs::ClientOptions(*creds));  // Same as one, but give you a chance to verify is the cred is working
3. // Set it up from a file, the content contains the project id, secret key, ext. Not recommended.  
   auto creds = gcs::oauth2::CreateServiceAccountCredentialsFromJsonFilePath(path)
    gcs::Client client(gcs::ClientOptions(*creds));
4. // Use GCE credentials for the instance's default service account.
   gcs::Client client(gcs::ClientOptions(gcs::oauth2::CreateComputeEngineCredentials()));

Reply:
I think they way you do is correct.